test/e2e/gateway_api_test.go

```diff
@@ -0,0 +1,212 @@
+
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/storage/names"
+
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	testHostname = "*.example.com"
+)
+
+var crdNames = []string{"gatewayclasses.gateway.networking.k8s.io",
+	"gateways.gateway.networking.k8s.io",
+	"httproutes.gateway.networking.k8s.io",
+	"referencegrants.gateway.networking.k8s.io",
+}
+
+// TestGatewayAPIResources tests that basic functions for Gateway API Custom Resources are functional.
+// It specifically verifies that when the GatewayAPI feature gate is enabled, then a user can
+// create a GatewayClass, Gateway, and HTTPRoute.
+func TestGatewayAPIResources(t *testing.T) {
+	t.Parallel()
+
+	// Get the cluster feature gate
+	featureGate := &configv1.FeatureGate{}
+	err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		name := types.NamespacedName{"", "cluster"}
+		if err := kclient.Get(context.TODO(), name, featureGate); err != nil {
+			t.Logf("failed to get feature gate %s: %v", gwapi.FeatureGateName, err)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to find feature gate: %v", err)
+	}
+
+	// Create a test namespace if needed
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: names.SimpleNameGenerator.GenerateName("test-e2e-gwapi-"),
+		},
+	}
+	if err := kclient.Create(context.TODO(), ns); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			t.Fatalf("error creating namespace: %v", err)
+		}
+	}
+	defer func() {
+		// Deleting the namespace also deletes the httproute
+		if err := kclient.Delete(context.TODO(), ns); err != nil {
+			t.Fatalf("failed to delete test namespace %s: %v", ns.Name, err)
+		}
+	}()
+
+	// Check if the feature gate is disabled, and if it is, make sure no GWAPI CRDs can be created
+	if !gwapi.FeatureIsEnabled(gwapi.FeatureGateName, featureGate) {
+		// First make sure nothing can happen with the Gateway API CRDs when the feature is disabled
+		ok, _, _ := testCrds(t, ns.Name, false)
+		if ok {
+			t.Fatalf("feature gate was not enabled, but gateway api objects could be created")
+		}
+		// Enable the feature gate for the rest of the test
+		featureGate.Spec.FeatureGateSelection.FeatureSet = configv1.CustomNoUpgrade
+		featureGate.Spec.FeatureGateSelection.CustomNoUpgrade = &configv1.CustomFeatureGates{
+			Enabled: []string{gwapi.FeatureGateName},
+		}
+		if err := kclient.Update(context.TODO(), featureGate); err != nil {
+			t.Fatalf("error enabling feature gate: %v", err)
+		}
+	}
+	// Make sure all the *.gateway.networking.k8s.io CRDs are available
+	for _, crdName := range crdNames {
+		err := getCrd(t, crdName)
+		if err != nil {
+			t.Fatalf("failed to find crd %s: %v", crdName, err)
+		}
+	}
```

Comment:
The two things I wanted to do were to 1) make sure you couldn't create the objects if the feature gate wasn't enabled and 2) make sure you could create the objects when the feature gate is enabled.  There may be a cleaner way to do that, but assertCRDExists doesn't test the creation of the object.

Reply:
I think I understand.  You're saying, "Make sure you cannot create Gateway API objects if the feature isn't enabled", and I'm saying, "Make sure the CRDs don't exist if the feature isn't enabled".  Really, what we want to verify is that cluster-ingress-operator doesn't create the CRDs if the feature isn't enabled, but that's impractical to verify directly.  Anyway, it makes more sense to me to check whether the CRDs are there, but if you prefer attempting creation, I can accept that.  

However, if you do continue with the current approach, `testCrds` is a tad non-obvious, and so is the test logic around it.  If you want to keep the current approach, I'd still encourage you to log unconditionally to simplify the logic, and maybe name the function something like `createCRDs` or even `createGatewayAPIResources`, if that isn't too verbose.  Hmm...  Or maybe change it into `assertCanCreateCRDs(t *testing.T, ns string, expectSuccess bool)`, and then it can still do the create but then it can do a `if expectSuccess { t.Fatal(err) }` in case of errors and do an `} else if !expectSuccess { t.Fatalf("expected create of thing to fail") }` when the `Create` doesn't return an error.